allow json callback customisation

petebacondarwin opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
feat

**What is the current behavior? (You can also link to an open issue here)**
hard-coded jsonp callback handling

**What is the new behavior (if this is a feature change)?**
overridable `$jsonpCallbacks` service that offers a place where customisation can be applied

**Does this PR introduce a breaking change?**
No

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

See the commits

petebacondarwin committed:
feat($jsonpCallbacks): new service to abstract how JSONP callbacks are handled

You can now override this service if you have specific requirements about
the behaviour and formatting of the JSON_CALLBACK that is sent to the server
for `$http.jsonp` requests.

petebacondarwin committed:
refact($http): use the $jsonpCallbacks service

Use the built-in service to handling callbacks to `$http.jsonp` requests.

Closes #3073

gkalpak commented on test/ng/httpBackendSpec.js:
```diff
@@ -310,25 +330,28 @@ describe('$httpBackend', function() {
 
 
     it('should clean up the callback and remove the script', function() {
+      spyOn($jsonpCallbacks, 'wasCalled').and.callThrough();
```
Is this needed ?

petebacondarwin replied:
Nope

gkalpak commented on src/ng/jsonpCallbacks.js:
```diff
@@ -0,0 +1,83 @@
+/**
+ * @ngdoc service
+ * @name $jsonpCallbacks
+ * @description
```
I'm not sure we are consistent with all services, but could add a `@requires $window`.

petebacondarwin replied:
Done

gkalpak commented on src/ng/jsonpCallbacks.js:
```diff
@@ -0,0 +1,83 @@
+    $window.angular.callbacks = {};
+    var callbackMap = {};
+
+    function createCalback(callbackId) {
```
_createCalback_  -->  _createCallback_

petebacondarwin replied:
Thanks

gkalpak commented on src/ng/jsonpCallbacks.js:
```diff
@@ -0,0 +1,83 @@
+       * in the JSONP response.
+       */
+      getResponse: function(callbackPath) {
+        var callback = callbackMap[callbackPath];
```
Nitpick: `return callbackMap[callbackPath].data` would be more consistent with `wasCalled`.

petebacondarwin replied:
Done

gkalpak commented on src/ng/jsonpCallbacks.js:
```diff
@@ -0,0 +1,83 @@
+       * callback that was passed in the request.
+       */
+      wasCalled: function(callbackPath) {
+        return callbackMap[callbackPath].called;
```
Since we are deleting the entries, do we need to safeguard against the `callbackPath` property not existing on `callbackMap` ?

petebacondarwin replied:
According to the way that I have written it in `$httpBackend` this should never happen.
If it does then we have a programming error somewhere, which should trigger an error so that we can know to fix it.

gkalpak commented on test/ng/jsonpCallbacksSpec.js:
```diff
@@ -0,0 +1,73 @@
+'use strict';
+
+describe('$jsonpCallbacks', function() {
+  var $mockWindow;
```
Not used.

petebacondarwin replied:
Thanks

gkalpak commented:
A few minor comments  - LGTM otherwise.
IE doesn't seem to share my opinion though :stuck_out_tongue:

petebacondarwin commented:
I fixed up IE... the problem was that occasionally, in the test, it was actually triggering the script callback at some point in the future, after the test had completed. So I put in a check that the status of the response is a 200 before trying to get hold of the response.

petebacondarwin committed:
feat($jsonpCallbacks): new service to abstract how JSONP callbacks are handled

You can now override this service if you have specific requirements about
the behaviour and formatting of the JSON_CALLBACK that is sent to the server
for `$http.jsonp` requests.

petebacondarwin committed:
refact($http): use the $jsonpCallbacks service

Use the built-in service to handling callbacks to `$http.jsonp` requests.

Closes #3073

petebacondarwin commented:
@gkalpak thanks for the review. PTAL

gkalpak commented on src/ng/httpBackend.js:
```diff
@@ -59,16 +59,11 @@ function createHttpBackend($browser, createXhr, $browserDefer, callbacks, rawDoc
-        callbacks[callbackId] = noop;
+      var callbackPath = callbacks.createCallback(url);
+      var jsonpDone = jsonpReq(url, callbackPath, function(status, text) {
+        var response = (status === 200) && callbacks.getResponse(callbackPath);
```
Minor: I would add a note that `status` is set in `jsonpReq()` and it can be either `-1`, `404` or `200`.
The first time I saw this, I thought: "What if it is in the (200, 300) range?"

gkalpak commented:
LGTM

petebacondarwin commented:
Thanks. I'll comment on the `status` codes and merge
